Userland/Libraries/LibWeb/DOM/Slottable.cpp

```diff
@@ -0,0 +1,205 @@
+/*
+ * Copyright (c) 2023, Tim Flynn <trflynn89@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <LibWeb/Bindings/MainThreadVM.h>
+#include <LibWeb/DOM/Element.h>
+#include <LibWeb/DOM/Node.h>
+#include <LibWeb/DOM/ShadowRoot.h>
+#include <LibWeb/DOM/Slottable.h>
+#include <LibWeb/DOM/Text.h>
+#include <LibWeb/HTML/HTMLSlotElement.h>
+
+namespace Web::DOM {
+
+SlottableMixin::~SlottableMixin() = default;
+
+void SlottableMixin::visit_edges(JS::Cell::Visitor& visitor)
+{
+    visitor.visit(m_assigned_slot);
+    visitor.visit(m_manual_slot_assignment);
+}
+
+// https://dom.spec.whatwg.org/#dom-slotable-assignedslot
+JS::GCPtr<HTML::HTMLSlotElement> SlottableMixin::assigned_slot()
+{
+    auto* node = dynamic_cast<DOM::Node*>(this);
+    VERIFY(node);
+
+    // The assignedSlot getter steps are to return the result of find a slot given this and with the open flag set.
+    return find_a_slot(node->as_slottable(), OpenFlag::Set);
+}
+
+JS::GCPtr<HTML::HTMLSlotElement> assigned_slot_for_node(JS::NonnullGCPtr<Node> node)
+{
+    if (!node->is_slottable())
+        return nullptr;
+
+    return node->as_slottable().visit([](auto const& slottable) {
+        return slottable->assigned_slot_internal();
+    });
+}
+
+// https://dom.spec.whatwg.org/#slotable-assigned
+bool is_an_assigned_slottable(JS::NonnullGCPtr<Node> node)
+{
+    if (!node->is_slottable())
+        return false;
+
+    // A slottable is assigned if its assigned slot is non-null.
+    return assigned_slot_for_node(node) != nullptr;
+}
+
+// https://dom.spec.whatwg.org/#find-a-slot
+JS::GCPtr<HTML::HTMLSlotElement> find_a_slot(Slottable const& slottable, OpenFlag open_flag)
+{
+    // 1. If slottable’s parent is null, then return null.
+    auto* parent = slottable.visit([](auto& node) { return node->parent_element(); });
+    if (!parent)
+        return nullptr;
+
+    // 2. Let shadow be slottable’s parent’s shadow root.
+    auto* shadow = parent->shadow_root_internal();
+
+    // 3. If shadow is null, then return null.
+    if (shadow == nullptr)
+        return nullptr;
+
+    // 4. If the open flag is set and shadow’s mode is not "open", then return null.
+    if (open_flag == OpenFlag::Set && shadow->mode() != Bindings::ShadowRootMode::Open)
+        return nullptr;
+
+    // 5. If shadow’s slot assignment is "manual", then return the slot in shadow’s descendants whose manually assigned
+    //    nodes contains slottable, if any; otherwise null.
+    if (shadow->slot_assignment() == Bindings::SlotAssignmentMode::Manual) {
+        JS::GCPtr<HTML::HTMLSlotElement> slot;
+
+        shadow->for_each_child_of_type<HTML::HTMLSlotElement>([&](auto& child) {
```

Comment:
"descendants" here means the entire subtree
https://dom.spec.whatwg.org/#concept-tree-descendant

```suggestion
        shadow->for_each_in_subtree_of_type<HTML::HTMLSlotElement>([&](auto& child) {
```